exchange/auction.go

```diff
@@ -88,61 +111,56 @@ func (a *auction) doCache(ctx context.Context, cache prebid_cache_client.Client,
 	for _, imp := range bidRequest.Imp {
 		expByImp[imp.ID] = imp.Exp
 	}
-	for _, topBidsPerImp := range a.winningBidsByBidder {
-		for _, topBidPerBidder := range topBidsPerImp {
-			impID := topBidPerBidder.bid.ImpID
-			isOverallWinner := a.winningBids[impID] == topBidPerBidder
-			if !includeBidderKeys && !isOverallWinner {
-				continue
+
+	/* bidders vs winners			*/
+	if targData.includeBidderKeys {
+		// cache banners and/or videos found in impsToBiddersTopBids map[string]map[openrtb_ext.BidderName]*pbsOrtbBid
+		for impID := range a.impsToBiddersTopBids {
+			for _, bid_to_cache := range a.impsToBiddersTopBids[impID] {
+				bids_to_cache_arr = append(bids_to_cache_arr, bid_to_cache)
 			}
-			var customCacheKey string
-			var catDur string
-			useCustomCacheKey := false
-			if competitiveExclusion && isOverallWinner {
-				// set custom cache key for winning bid when competitive exclusion applies
-				catDur = bidCategory[topBidPerBidder.bid.ID]
-				if len(catDur) > 0 {
-					customCacheKey = fmt.Sprintf("%s_%s", catDur, hbCacheID)
-					useCustomCacheKey = true
-				}
+		}
+	} else if targData.includeWinners {
+		// cache banners and/or videos found in impsToTopBids map[string]*pbsOrtbBid
+		for _, bid_to_cache := range a.impsToTopBids {
+			bids_to_cache_arr = append(bids_to_cache_arr, bid_to_cache)
+		}
+	}
+
+	for _, bid_to_cache := range bids_to_cache_arr {
+		/* banners vs videos 			*/
+		if targData.includeCacheBids { //banner
+			if jsonBytes, err := json.Marshal(bid_to_cache.bid); err == nil {
+				toCache = append(toCache, prebid_cache_client.Cacheable{
+					Type:       prebid_cache_client.TypeJSON,
+					Data:       jsonBytes,
+					TTLSeconds: cacheTTL(expByImp[bid_to_cache.bid.ImpID], bid_to_cache.bid.Exp, defTTL(bid_to_cache.bidType, defaultTTLs), ttlBuffer),
+				})
+				bidIndices[len(toCache)-1] = bid_to_cache.bid
+			} else {
+				errs = append(errs, err)
 			}
-			if bids {
-				if jsonBytes, err := json.Marshal(topBidPerBidder.bid); err == nil {
-					if useCustomCacheKey {
-						// not allowed if bids is true; log error and cache normally
-						errs = append(errs, errors.New("cannot use custom cache key for non-vast bids"))
-					}
+		}
+		if targData.includeCacheVast && bid_to_cache.bidType == openrtb_ext.BidTypeVideo { //video
+			if jsonBytes, err := json.Marshal(makeVAST(bid_to_cache.bid)); err == nil {
+				catDur, ok := bidCategory[a.impsToTopBids[bid_to_cache.bid.ImpID].bid.ID]
```

Comment:
It's probably more efficient to assign catDur only once it's known to be needed.  So, probably better to check if it's a winning bid first, then assign catDur.

Reply:
Agree, done.